### ISSUE TITLE
Phase 2: Sequential LoRA Fine-tuning Implementation

### ISSUE THREAD

**Issue by zhc1212** [zhc1212/SVD-LLM#2]
## Context

SVD-LLM(W) (Phase 1) 已完成：白化 + SVD 截断压缩，结果优于论文。Phase 2 实现 Sequential LoRA 微调（论文的 SVD-LLM），在压缩后模型上用 Alpaca 50K 做两阶段 LoRA 微调恢复精度。

## 核心流程

```
加载原始模型 → Stage A 压缩(得到 CompressedLinear) → Stage B LoRA 微调 → 合并 → 保存
```

Stage B 分两个阶段：
1. **Stage 1 (LoRA_u)**: 冻结 W'_v (`.first`)，LoRA 微调 W'_u (`.second`)，合并 LoRA
2. **Stage 2 (LoRA_v)**: 冻结 W'_u (`.second`)，LoRA 微调 W'_v (`.first`)，合并 LoRA

> 注：已保存的压缩模型是合并后的 nn.Linear 格式，需要重新跑 Stage A 压缩来恢复 CompressedLinear 结构。Stage A 约 10 分钟，可接受。

### 1. `src/data/alpaca.py` — Alpaca 数据集加载

- 加载 `yahma/alpaca-cleaned`
- 标准 Alpaca instruction prompt 模板
- Tokenize + label masking（prompt 部分标 -100，只在 response 上计算 loss）
- 接口：`get_alpaca_dataset(tokenizer, max_length=256, seed=42) → Dataset`

### 3. `src/finetune/sequential_lora.py` — 核心 Sequential LoRA 逻辑

关键函数：
- `finetune_sequential_lora(model, tokenizer, ratio, ...)` — 主入口
  1. 调用 `compress_model_whitening_only()` 重新压缩得到 CompressedLinear
  2. 加载 Alpaca 数据集
  3. Stage 1: `LoraConfig(target_modules=["second"])` + Trainer 训练 + `merge_and_unload()`
  4. Stage 2: `LoraConfig(target_modules=["first"])` + Trainer 训练 + `merge_and_unload()`
  5. 返回 CompressedLinear 形态的模型

- `_run_lora_stage(model, dataset, target_modules, ...)` — 单阶段 LoRA
  - 用 PEFT `get_peft_model` 包装
  - HuggingFace `Trainer` 训练
  - `merge_and_unload()` 合并回去

### 4. `scripts/finetune.py` — CLI 入口

```bash
python scripts/finetune.py \
    --model_path /path/to/original/llama-7b \
    --ratio 0.2 \
    --save_path outputs/llama7b_svd_llm_20 \
    --lora_r 32 --lora_alpha 64 --lr 2e-4 --epochs 1
```

读取压缩参数 → 调用 `finetune_sequential_lora()` → `merge_compressed_model()` → 保存

### 5. `scripts/run_svd_llm_experiments.sh` — 实验脚本

4 个压缩率 (20/40/60/80%) 串行运行 finetune + eval。

## LoRA 超参数

| 参数 | 值 | 说明 |
|------|-----|------|
| lora_r | 32 | LoRA 秩 |
| lora_alpha | 64 | alpha/r=2 |
| lora_dropout | 0.05 | |
| lr | 2e-4 | |
| epochs | 1 | 每阶段 1 epoch |
| batch_size | 4 | per-device |
| grad_accum | 4 | 有效 batch=16 |
| max_length | 256 | Alpaca 序列长度 |
| gradient_checkpointing | True | 省显存 |
| fp16 | True | |

## 显存估算 (单卡 A800 80GB)

- Stage A 重压缩：~40GB（临时，跑完释放）
- Stage B LoRA 训练：~18GB（模型 13GB + LoRA + 优化器 + 激活）
- 单卡完全够用

## 依赖

- `peft>=0.6.0`（已安装）
- `datasets`（已安装）
- 现有模块：`compress_model_whitening_only`, `merge_compressed_model`, `get_calibration_data`, `load_model`

## 实现顺序

1. `src/data/alpaca.py`
2. `src/finetune/__init__.py` + `src/finetune/sequential_lora.py`
3. `scripts/finetune.py`
4. `scripts/run_svd_llm_experiments.sh`
5. 跑实验，更新 README 结果

## 验证

1. 小规模冒烟测试：2 层、10 samples，验证 LoRA 应用/合并/保存正常
2. 完整实验：4 个压缩率 × (PPL + downstream) 评估
3. 对比论文 SVD-LLM 列填入 README
